elector/elector.go

```diff
@@ -0,0 +1,286 @@
+			isLeader, leaderNode, err := determineLeader(lockPath, children)
+			if err != nil {
+				return errors.Wrap(err, "could not determine leader")
+			}
+			leaderIdent, err := e.getIdentFromNode(leaderNode)
+			if err != nil {
+				return errors.Wrap(err, "could not get ident from node")
+			}
+			e.setLeaderIdent(leaderIdent)
+			e.sendLeader(isLeader, firstLeaderUpdate)
+			firstLeaderUpdate = false
+			return nil
+		}
+
+		children, _, childEvents, err := e.conn.ChildrenW(e.basePath)
+		if err != nil {
+			return errors.Wrap(err, "could not get children")
+		}
+		if err := updateFunc(children); err != nil {
+			return err
+		}
+		for {
+			select {
+			case zkEvent := <-zkEvents:
+				if zkEvent.Err != nil {
+					return zkEvent.Err
+				}
+				switch zkEvent.State {
+				case zk.StateExpired, zk.StateAuthFailed, zk.StateDisconnected, zk.StateUnknown:
+					return fmt.Errorf("invalid ZK state: %v", zkEvent.State)
+				}
+			case <-childEvents:
+				children, _, childEvents, err = e.conn.ChildrenW(e.basePath)
+				if err != nil {
+					return errors.Wrap(err, "could not get children")
+				}
+				if err := updateFunc(children); err != nil {
+					return err
+				}
+			}
+		}
+	}()
+	// the elector errored out unexpectedly. send an error to the client.
+	e.sendErr(err)
+}
+
+// getIdentFromNode fetches the znode data from the specified node and returns
+// it as a string
+func (e *Elector) getIdentFromNode(node string) (ident string, err error) {
+	nodePath := path.Join(e.basePath, node)
+	b, _, err := e.conn.Get(nodePath)
+	return string(b), err
+}
+
+// sendLeader sends a leader event on the events chan.  if the previous
+// leader state did not change, it will not send the event unless the
+// force parameter is true.
+func (e *Elector) sendLeader(leader bool, force bool) {
+	e.mut.Lock()
+	isLeader := e.isLeader
+	e.mut.Unlock()
+	if isLeader == leader && !force {
+		return
+	}
+	e.mut.Lock()
+	e.isLeader = leader
+	e.mut.Unlock()
+	e.events <- Event{Leader: leader}
+}
+
+// setLeaderIdent safely sets the current leader ident on the elector. This
+// allows it to be queried for the current leader ident without having to
+// consult zookeeper every time.
+func (e *Elector) setLeaderIdent(ident string) {
+	e.mut.Lock()
+	defer e.mut.Unlock()
+	e.leaderIdent = ident
+}
+
+// sendErr sends an error event on the events chan.
+func (e *Elector) sendErr(err error) {
+	e.events <- Event{Err: err}
+}
+
+// sorted children sequences converts the children to sequence parts, and
+// then returns the sorted sequences, along with a lookup map of sequence
+// to nodes
+func sortedChildrenSequences(children []string) (sorted []int, lookup map[int]string, err error) {
+	sorted = make([]int, len(children))
+	lookup = make(map[int]string)
+	for i, child := range children {
+		seq, err := sequencePart(child)
+		if err != nil {
+			return nil, nil, err
+		}
+		sorted[i] = seq
+		lookup[seq] = child
+	}
+	return sorted, lookup, nil
```

Comment:
I was expecting some kind of call to a sorting algorithm to actually sort the `sorted` sequences. I don't see one. Is that intentional?

Reply:
ah .. looks like the call to `sort.Slice` happens later (determineLeader). any reason why it's not here instead?

Reply:
Ugh, thanks. Yeah I'll fix this up, thanks.